src/pubsub/ua_pubsub_reader.c

```diff
@@ -1839,4 +1722,270 @@ UA_PubSubConnection_processNetworkMessage(UA_Server *server, UA_PubSubConnection
+
+    UA_DataSetMessage_freeDecodedPayload(
+        dataSetReader->bufferedMessage.nm->payload.dataSetPayload.dataSetMessages);
+
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+    useNormalAlloc();
+#endif /* UA_ENABLE_PUBSUB_BUFMALLOC */
+
+    /* Minimum ethernet packet size is 64 bytes where the header size is 14
+     * bytes and FCS size is 4 bytes so remaining minimum payload size of
+     * ethernet packet is 46 bytes */
+    /* TODO: Need to handle padding bytes for UDP */
+    if(((*currentPosition - previousPosition) < MIN_PAYLOAD_SIZE_ETHERNET) &&
+       (strncmp((const char *)connection->config->transportProfileUri.data,
+                "http://opcfoundation.org/UA-Profile/Transport/pubsub-eth-uadp",
+                (size_t)(connection->config->transportProfileUri.length)) == 0)) {
+        paddingBytes =
+            (MIN_PAYLOAD_SIZE_ETHERNET - ((*currentPosition) - previousPosition));
+        (*currentPosition) += paddingBytes; /* During multiple receive, move the position
+                                               to handle padding bytes */
+    }
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+receiveBufferedNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
+                              UA_PubSubConnection *connection) {
+    UA_ByteString buffer;
+    buffer.length = RECEIVE_MSG_BUFFER_SIZE;
+    buffer.data = ReceiveMsgBuffer;
+
+    UA_StatusCode rv = connection->channel->receive(
+        connection->channel, &buffer, NULL, readerGroup->config.timeout);
+
+    // TODO attention: here rv is ok if UA_STATUSCODE_GOOD != rv
+    UA_CHECK_WARN(!UA_StatusCode_isBad(rv), return rv,
+                  &server->config.logger, UA_LOGCATEGORY_SERVER,
+                  "SubscribeCallback(): Connection receive failed!");
+    if(buffer.length > 0) {
+        size_t currentPosition = 0;
+        size_t previousPosition = 0;
+        if (readerGroup->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
+            do {
+                rv = decodeAndProcessNetworkMessageRT(server, readerGroup, connection,
+                                                      previousPosition, &buffer,
+                                                      &currentPosition);
+                UA_CHECK_STATUS_WARN(
+                    rv, (void) 0, &server->config.logger, UA_LOGCATEGORY_SERVER,
+                    "SubscribeCallback(): receive message failed");
+
+                previousPosition = currentPosition;
+
+            } while((buffer.length) > currentPosition);
+        } else {
+            UA_LOG_DEBUG(&server->config.logger, UA_LOGCATEGORY_USERLAND,
+                         "Message received:");
+            do {
+                rv = decodeAndProcessNetworkMessage(server, readerGroup, connection,
+                                                    previousPosition, &buffer,
+                                                    &currentPosition);
+
+                // TODO: check what actions are to be done on error.. nothing?
```

Comment:
Yes. Break the loop on error and return rv.

Reply:
That also was my initial thought, but it broke some tests.